linux kexec: skip kernel / initrd copying when source is a regular file.

10000TB opened the pull request:
...and under tmpfs.

One caveat discussed is original file may be opened
for writting, which kexec file load does not like.
This change does not proactively prevent that from
happening within linux.go, but now allow it to fail
in that case.

One thing I can do is write, in golang, lsof, and matching
/proc/\*/fd/\* sym linked file against original file name
and see if there is any currently opening it, and throw
an error to kexec load caller, but that is not giving
any advantage over let kexec file load propogating up
the error. One benefit being we can do sth custom when
that happens, say proactively purging / closing them.
That just lead us into another hole of mess. So I prefer
to "Leap Before You Look" here for simplicity.

Signed-off-by: David Hu <xuehaohu@google.com>

hugelgupf commented:
I don't think this should be a task in the boot image, it should be a task
on the caller.

boot.LinuxImage uses io.ReaderAt, which you cannot just assume to always
have a file path. You can only make that decision at the place where the
boot.LinuxImage is actually created, and I'd suggest wrapping another
menu.Entry I guess? I don't know, that seems a bit obtuse. Not sure I like
either option.

On Mon, Aug 30, 2021, 10:27 Xuehao Hu ***@***.***> wrote:

> @10000TB <https://github.com/10000TB> requested review from
> @u-root/contributors on: #2112
> <https://github.com/u-root/u-root/pull/2112> linux kexec: add option
> allow deleting original image before kexec..
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/u-root/u-root/pull/2112#event-5227655029>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPG3EWRS4TCCVUZDY5FA7DT7O5PTANCNFSM5DCJNAAA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

fotonick commented:
@hugelgupf n00b thought: In this case and others, it seems like it could be desirable to separate a `kexec_load` and `kexec_exec`—I can imagine multiple types of cleanup/deinitialization you might want to do before actually handing off to the next OS but after staging the image. Does that make any sense?

hugelgupf commented:
Load and exec are already separated

On Mon, Aug 30, 2021, 13:30 Nickolas Fotopoulos ***@***.***>
wrote:

> @hugelgupf <https://github.com/hugelgupf> n00b thought: In this case and
> others, it seems like it could be desirable to separate a kexec_load and
> kexec_exec—I can imagine multiple types of cleanup/deinitialization you
> might want to do before actually handing off to the next OS but after
> staging the image. Does that make any sense?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/u-root/u-root/pull/2112#issuecomment-908671752>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPG3ERLGZLFBDGYQEFZ42LT7PS7DANCNFSM5DCJNAAA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

hugelgupf commented:
But to expand on that:

Each menu.Entry has a Load and Exec. Ultimately, load is called as part of
the menu here:
https://github.com/u-root/u-root/blob/9a482f90d8eb33b97e019c0e8b74bfc40f8aaa50/pkg/boot/bootcmd/bootcmd.go#L33

And exec here:
https://github.com/u-root/u-root/blob/9a482f90d8eb33b97e019c0e8b74bfc40f8aaa50/pkg/boot/bootcmd/bootcmd.go#L49

However, the cleanup that happens in between is system cleanup - you want
to clean up things that are independent from menu entries there.

Stuff dependent on which entry was chosen should probably be cleaned up as
part of Load or Exec.

I think overall I do agree with David's approach here

On Mon, Aug 30, 2021, 14:13 Chris Koch ***@***.***> wrote:

> Load and exec are already separated
>
> On Mon, Aug 30, 2021, 13:30 Nickolas Fotopoulos ***@***.***>
> wrote:
>
>> @hugelgupf <https://github.com/hugelgupf> n00b thought: In this case and
>> others, it seems like it could be desirable to separate a kexec_load and
>> kexec_exec—I can imagine multiple types of cleanup/deinitialization you
>> might want to do before actually handing off to the next OS but after
>> staging the image. Does that make any sense?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/u-root/u-root/pull/2112#issuecomment-908671752>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAPG3ERLGZLFBDGYQEFZ42LT7PS7DANCNFSM5DCJNAAA>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>>
>

10000TB commented:
another bummer of moving the cleanup to between load and the exec, is that cleanup must happen before load, as load actually takes the tmpfs kernel and initrd into memory.

another alternative considered is let cleanup logic be passed in as a function by caller (it knows if it is a local file), either through func parameter, or a new field in Linux image, both of them looked weird...but, wherever we move the copytoFile to, this cleanup can be moved there.

It will simply fail (&proceed further) when original "file" is read on the fly than have been stored on extract_dir, but that is  not worse as no need to free it as no duplicate copy.

10000TB commented:
the basic arm integration is flaky -- previously failing, now ok

10000TB commented:
ping: need an approval

10000TB commented:
discussed offline -- skipping file copy when source is a regular file is preferred.

hugelgupf commented on pkg/boot/linux.go:
```diff
@@ -62,7 +63,36 @@ func (li *LinuxImage) String() string {
+func copyToFile(r io.Reader, name string) (*os.File, error) {
+	// If source is a regular file in tmpfs, simply re-use that than copy.
+	if _, err := os.Stat(name); err == nil { // Name is the path.
+		if r, _ := mount.IsTmpfs(name); r {
```
this is brittle. Nothing guarantees that an *os.File.Name() is actually a valid path. Anybody can call os.NewFile with a random FD and a random string, and that FD can turn out to be anything, and so can that string. I much prefer what I showed you internally, where the caller **explicitly** passes a type that means copying will be skipped, and that type enforces that the file that is being created is a tmpfs file.

10000TB replied:
That is a good point... newFile only simply pass the name to  golang `file` struct -- and `Open` / `OpenFile` on the name would break

Just to understand a bit more before change approach, is that a real use case ? or possible theoretical scenario to rule out for completeness.

10000TB replied:
(ping)

rjoleary replied:
I might be missing something, where is *os.File.Name() called?

hugelgupf replied:
`copyToFile(kernel, stringer(li.Kernel))` == `copyToFile(kernel, kernel.Name())` it's called by the caller.

hugelgupf replied:
people don't intentionally pass the wrong path. they intentionally pass no path when there isn't one, e.g. in these cases https://github.com/u-root/u-root/search?q=os.NewFile

rjoleary replied:
Won't this degrade gracefully? The `os.Stat(name)` will fail is the name is blank and you fall back to the code that copies.

rminnich replied:
THis is why fstat exists

rjoleary replied:
Can fstat tell you if the fd is on a tmpfs?

rjoleary replied:
I think you can also use `readlink /proc/self/fd/#`, but Linux specific.

10000TB replied:
the file descriptor is not available here, as the kernel and initramfs is passed in via `io.ReaderAt`. 

if original caller passes in a file via `newFile` on a given descriptor, and the `name` given is not legit, I think that is a dead end, as `fstat` system call also depends on a valid path.

but note that this is now within 

```
if fi, err := os.Stat(name); err == nil && fi.Mode().IsRegular() {
   ...
}
```
it won't enter this branch, for the specific case -- the addition of this PR should be a no-op for that specific case, as it will fall back to copyToFile. The question is if that is sth we need to also to avoid copy ? -- I tend think we should copy, as it sounds like original file is probably not in tmpfs

hugelgupf replied:
I'm not really comfortable with the potential for confusion or the potential for misusing this interface. And I'm not comfortable with that discomfort being ignored.

If stringer(li.Kernel) is e.g. a vfile.File, but the vfile.File.Name() happens to be an actual file path on the real file system, the stat will succeed, and we open and kexec the wrong file here.

If between the time that the file was opened and the file is being used here the file has been changed on the file system, this is different. I'm not comfortable with this solution.

hugelgupf replied:
If stringer(li.Kernel) is a random file we download from the internet, but (fmt.Stringer).String() happens to return a valid local file path, we'll just open that file and kexec that. That's not cool either.

10000TB replied:
To add another example: Even when File passed in is legit, fits to avoid copy (regular file, on tmpfs), caller may still do sth to sabotage the file to be loaded, e.g the file maybe opened for additional operation (for whatever reason, who knows), and that corrupted it - one temptation is to write code here (in kexec) to prevent that  from happening (note, the operation could just finish before kexec file load call, so it won't confuse with there is ps still editing it)

I don't think we will ever be able to add checks to prevent all kinds of those things. However, we can't  check nothing here neither. There need a balance.

I am happy to incorporate additional checks though, similar to the nature of "source being regular file", "source in tmpfs", etc.

fotonick replied:
IIUC, the latest conversation is about TOCTOU attacks. There are some general techniques around holding onto the same file handle for the duration of your work or comparing inodes to see that multiple stats/opens refer to the same file. https://wiki.sei.cmu.edu/confluence/display/c/FIO45-C.+Avoid+TOCTOU+race+conditions+while+accessing+files

colinmitchell-ggl replied:
My understanding is there is just a lot of complexity in correctly automating the should_copy logic. For simplicity, can this be pushed to the caller who should have a better idea where their image is? The copy could hinge on an i.SkipCopy

10000TB replied:
re: TOCTOU -- that is a good point, re-opening with `name` creates a window that the same file may be swapped. Albeit that window is very small... updated the code a bit, to casting reader back to a `os.File` (a bad assumption?) and re-use the same logic to check tmpfs, etc. That saves a re-opening, should also get rid of other concerns.

re: skip copy -- yea, that is a good one, we could do that -- but at caller side, we still need to check if the directory is on tmpfs or not; but it for sure knows if it operates on a local `os.File`, or other types of file sources

re: interface -- I prefer we reserve this as last option, as this is one-use interface, and I find that a slightly over kill

rjoleary commented on pkg/boot/linux.go:
```diff
@@ -62,7 +63,36 @@ func (li *LinuxImage) String() string {
 
-func copyToFile(r io.Reader) (*os.File, error) {
+func copyToFile(r io.Reader, name string) (*os.File, error) {
+	// If source is a regular file in tmpfs, simply re-use that than copy.
```
I think I'm missing something. Where's the check that it's a regular file?

10000TB replied:
My understanding is that if `os.Stat` can tell the path exist from file system, it is a file -- regular in a sense it exists, other than need to read / stream from other types, which looks like could be "os.File could be a socket, a pipe or some other strange thing" (I don't know where we would run into such case, realistically)

rjoleary replied:
I think the terminology is slightly off. You can stat fifos and sockets like:

```
~$ mkfifo f
~$ stat f
  File: f
  Size: 0         	Blocks: 0          IO Block: 4096   fifo
...
```

10000TB replied:
You are right -- the code currently is missing an additional check, updated the code to be:

```
if fi, err := os.Stat(name); err == nil && fi.Mode().IsRegular() {
   ...
}
```

rjoleary commented on pkg/boot/linux.go:
```diff
@@ -62,7 +63,36 @@ func (li *LinuxImage) String() string {
+			//
+			// Same file maybe opened by another process for writing
+			// which is theoretically possible, but may not need to be
+			// paranoi on.
```
```suggestion
			// paranoid on.
```

10000TB replied:
Done

10000TB commented on pkg/mount/magic.go:
```diff
@@ -11,10 +12,16 @@ import (
 
+// Copies from https://github.com/torvalds/linux/blob/master/include/uapi/linux/magic.h.
+const (
+	LINUX_TMPFS_MAGIC = 0x01021994
```
👍  Done.

10000TB commented on pkg/boot/linux.go:
```diff
@@ -137,5 +167,6 @@ func (li *LinuxImage) Load(verbose bool) error {
 		}
 		log.Printf("Command line: %s", li.Cmdline)
 	}
+
```
Done.

10000TB commented on pkg/boot/linux.go:
```diff
@@ -62,7 +63,36 @@ func (li *LinuxImage) String() string {
 }
 
-func copyToFile(r io.Reader) (*os.File, error) {
+func copyToFile(r io.Reader, name string) (*os.File, error) {
```
Sure - Done.

10000TB commented on pkg/mount/magic.go:
```diff
@@ -215,3 +222,12 @@ func FSFromBlock(n string) (fs string, flags uintptr, err error) {
 }
+
+// IsTmpfs tells if the file path given is under a tmpfs.
+func IsTmpfs(path string) (bool, error) {
```
Done.

10000TB commented:
The test passed locally when run as root -- but failed in CI -- I'll look a bit 

```
xuehaohu@worldpeace:~/go/src/github.com/u-root/u-root$ GO111MODULE=off sudo /usr/lib/google-golang/bin/go test -v -run TestIsTmpfs ./pkg/mount/...
=== RUN   TestIsTmpfs
--- PASS: TestIsTmpfs (0.00s)
PASS
ok  	github.com/u-root/u-root/pkg/mount	(cached)
testing: warning: no tests to run
PASS
ok  	github.com/u-root/u-root/pkg/mount/block	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/u-root/u-root/pkg/mount/gpt	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/u-root/u-root/pkg/mount/loop	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/u-root/u-root/pkg/mount/mtd	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/u-root/u-root/pkg/mount/scuzz	(cached) [no tests to run]
```

rjoleary reviewed (approved):
Besides the test tailing, this looks fine to me.

10000TB commented:
submit this now, as all major concerns are addressed and there is an approval.

re-posting some final notes 
------

```
1. addressed concerns raised around TOCTOU -- type casting the file back to os.File, and re-use that than re-opening by Name() eliminates a potential for bad actors to delete original file, and fake a new one under the same name.
2. addressed concern raised on stringer(..) not reliable -- type casting it back to os.File, eliminates the need to re-open by Name().
3. also addressed a other suggestions - e.g. introducing a flag to skip -- this is good one, I can still add it, but left it as is for now, since we auto detect it already, and using the flag would still mean we need to make sure caller to do the check for files are under tmpfs (it for sure knows it is a local regular file); defining new type / interface to file type matching -- My view is that this can be cautioned against, as the type is going to be one-off, and not versatile, so prefer less code with the approach in the PR.
4. I can still go back to the approach @hugelgupf  has been suggesting, if everyone agrees that is better than current form -- let me know.
5. Casting it as os.File limit caller from pass custom type as kernel / initramfs into boot entry (e.g. vfile ?). But that can be easily improved to make it also work for custom types when needed. I can make some change towards that also. I don't see that is needed for now.
```

10000TB commented:
Looks like I cant just "Rebase and merge" due to 

```
This branch cannot be rebased due to conflicts
Rebasing the commits of this branch on top of the base branch cannot be performed automatically due to conflicts encountered while reapplying the individual commits from the head branch.
```

let me try force push and merge from master again -- to resolve conflict
